include/dgl/graph_interface.h

```diff
@@ -353,6 +345,32 @@ class GraphInterface {
    * \return a vector of IdArrays.
    */
   virtual std::vector<IdArray> GetAdj(bool transpose, const std::string &fmt) const = 0;
+
+  static constexpr const char* _type_key = "graph.Graph";
+  DGL_DECLARE_OBJECT_TYPE_INFO(GraphInterface, runtime::Object);
+};
+
+/*! \brief Base graph reference */
+class GraphRef : public runtime::ObjectRef {
```

Comment:
What's the difference between GraphRef/ GraphPtr/ GraphInterface?

Reply:
`GraphInterface` is an interface class. GraphPtr is `std::shared_ptr<GraphInterface>`. GraphRef is equivalent to GraphPtr except that it can be used in container. The general practice I found useful is that:

* Use shared pointer (i.e, GraphPtr) in C++ implementation.
* When returning GraphPtr to python, wrap it as GraphRef. Similarly, when getting the argument from python, also use GraphRef.
```c++
DGL_REGISTER_GLOBAL("SomeCAPI")
.set_body([] (DGLArgs args, DGLRetValue *rv) {
  GraphRef g = args[0];      // use graph ref here
  GraphPtr sptr = g.sptr();  // get the shared pointer
  GraphPtr retptr = sptr->SomeCall();   // C++ functions all work on GraphPtr
  *rv = GraphRef(retptr);    // return graph ref
});
```

Side notes:
* `GraphRef` also implements the pointer operator `->`, so you could use it just like a `GraphPtr`.
* Returning `GraphPtr` to python is also allowed. It is the same as returning GraphRef. However, I suggest we keep it consistent.

Reply:
Thanks for super clear explanation. We could add it to our developer guides also.